posthog/auth.py

```diff
@@ -175,6 +175,77 @@ def authenticate_header(cls, request) -> str:
         return cls.keyword
 
 
+class ProjectSecretAPIKeyAuthentication(authentication.BaseAuthentication):
+    """
+    Authenticates using a project secret API key. Unlike a personal API key, this is not associated with a user
+    and should only be used for local_evaluation requests. When authenticated, this returns a "synthetic"
+    ProjectSecretAPIKeyUser object that has the team set. This allows us to use the existing permissioning system
+    for local_evaluation requests.
+
+    Only the first key candidate found in the request is tried, and the order is:
+    1. Request Authorization header of type Bearer.
+    2. Request body.
+    3. Request query string.
+    """
+
+    @classmethod
+    def find_secret_api_token(
+        cls,
+        request: Union[HttpRequest, Request],
+    ) -> Optional[str]:
+        """Try to find project secret API key in request and return it"""
+        if "HTTP_AUTHORIZATION" in request.META:
+            authorization_match = re.match(rf"^Bearer\s+(phs_.+)$", request.META["HTTP_AUTHORIZATION"])
+            if authorization_match:
+                return authorization_match.group(1).strip()
+
+        # Wrap HttpRequest in DRF Request if needed
+        if not isinstance(request, Request):
+            request = Request(request)
+
+        data = request.data
+
+        if data and "secret_api_key" in data:
+            return data["secret_api_key"]
+
+        if "secret_api_key" in request.GET:
+            return request.GET["secret_api_key"]
+
+        return None
+
+    def authenticate(self, request: Union[HttpRequest, Request]) -> Optional[tuple[Any, None]]:
+        secret_api_token = self.find_secret_api_token(request)
+
+        if not secret_api_token:
+            return None
+
+        # get the team from the secret api key
+        try:
+            Team = apps.get_model(app_label="posthog", model_name="Team")
+            team = Team.objects.get_team_from_cache_or_secret_api_token(secret_api_token)
+
+            if team is None:
+                return None
+
+            # Secret api keys are not associated with a user, so we create a ProjectSecretAPIKeyUser
+            # and attach the team. The team is the important part here.
+            return (ProjectSecretAPIKeyUser(team), None)
+        except Team.DoesNotExist:
+            return None
+
+
+class ProjectSecretAPIKeyUser:
+    """
+    A "synthetic" user object returned by the ProjectSecretAPIKeyAuthentication when authenticating with a project secret API key.
+    """
```

Comment:
Something to consider: I'm also in the middle of porting this endpoint to Rust. Once it's moved to rust, we won't need this hack any longer.